text/3573-is.md

```diff
@@ -0,0 +1,299 @@
+```
+
+The condition using `is` is a natural translation from the `is_some_and`
+method, whereas the if-let construction requires reversing the binding of `v`
+and the expression producing the option. This seems sufficiently cumbersome in
+some cases that the absence of `is` would motivate continued use and
+development of helper methods.
+
+# Guide-level explanation
+
+Rust provides an `is` operator, which can be used in any expression:
+`EXPR is PATTERN`
+
+This operator tests if the value of `EXPR` matches the specified `PATTERN`; see
+<https://doc.rust-lang.org/reference/patterns.html> for details on patterns.
+
+If the `EXPR` matches the `PATTERN`, the `is` expression evaluates to `true`,
+and additionally binds any bindings specified in `PATTERN` in the current scope
+for code subsequently executed along the path where the `is` expression
+is known to be `true`.
+
+For example:
+
+```rust
+if an_option is Some(x) && x > 3 {
+    println!("{x}");
+}
+```
+
+The bindings in the pattern are not bound along any code path potentially
+reachable where the expression did not match:
+
+```rust
+if (an_option is Some(x) && x > 3) || (more_conditions /* x is not bound here*/) {
+    // x is not bound here
+} else {
+    // x is not bound here
+}
+// x is not bound here
+```
+
+The pattern may use alternation (within parentheses), but must have the same
+bindings in every alternative:
+
+```rust
+if color is (RGB(r, g, b) | RGBA(r, g, b, _)) && r == b && g < 10 {
+    println!("condition met")
+}
+
+// ERROR: `a` is not bound in all alternatives of the pattern
+if color is (RGB(r, g, b) | RGBA(r, g, b, a)) && r == b && g < 10 {
+    println!("condition met")
+}
+```
+
+`is` may appear anywhere a boolean expression is accepted:
+
+```rust
+func(x is Some(y) && y > 3);
+```
+
+The `is` operator may not appear as a statement, because the bindings won't be
+usable after the end of the statement, and because the boolean return value is
+ignored. The compiler will issue a deny-by-default lint in that case, and
+suggest using `let` if the user wants to bind a pattern for the rest of the
+scope.
+
+```rust
+// deny-by-default lint: the binding `x` isn't usable after the statement, and the value of `is` is ignored.
+// Suggestion: use `let` to introduce a binding in this scope: `let x = an_expression();`.
+an_expression() is x;
+```
+
+# Reference-level explanation
+
+Add a new [operator
+expression](https://doc.rust-lang.org/reference/expressions/operator-expr.html),
+`IsExpression`:
+
+> **<sup>Syntax</sup>**\
+> _IsExpression_ :\
+> &nbsp;&nbsp; _Expression_ `is` _PatternNoTopAlt_
+
+Add `is` to the [operator
+precedence](https://doc.rust-lang.org/reference/expressions.html#expression-precedence)
+table, at the same precedence level as `==`, and likewise non-associative
+(requiring parentheses).
+
+Detect `is` appearing as a top-level statement and produce an error, with a
+rustfix suggestion to use `let` instead.
```

Comment:
This Reference-Level Explanation is too short to explain how `is` actually works, especially compared with the Guide-Level Explanation. 

I would like to see how that ***and additionally binds any bindings specified in `PATTERN` in the current scope for code subsequently executed along the path where the `is` expression is known to be `true`*** works, not through just a small number of examples, but the actual algorithm over all possible combinations of `$:expr`s, for instance the Guide-Level Explanation is not clear whether this is valid or not:

```rust
match x {
    Some(y) if y is Some(z) => (z is Some(w)).then(|| w + 1),
    _ => ..,
}
```

Reply:
@kennytm Valid. So, you're suggesting a systematic look at every existing statement type and which parts of the statement the binding is valid in? That seems entirely reasonable.

Reply:
> So, you're suggesting a systematic look at every existing statement type and which parts of the statement the binding is valid in?

Yes. Expressions too (well basically all statements except `let` and items are expressions), e.g. stuff like `((x is Some(z)) || (y is Some(z)) && z > 0`.